### PR TITLE
Added Q.any(promisesArray) method.

### PR DESCRIPTION
**Pull request by vergara**
Returns a promise fulfilled with the value of the first resolved promise in promisesArray. If all promises in promisesArray are rejected, it returns a rejected promise.

### CODE REVIEW

**Comment by vergara** [comment]
@kriskowal 
Updated.

I was not able to re-open the previous PR.

Do you have any feedback regarding the tests?

File: q.js [hunk 1]
```diff
@@ -1540,6 +1540,56 @@ Promise.prototype.all = function () {
...
+ * Returns the first resolved promise of an array. Prior rejected promises are
+ * ignored.  Rejects only if all promises are rejected.
+ * @param {Array*} an array containing values or promises for values
+ * @returns a promise fulfilled with the value of the first resolved promise,
+ * or a rejected promise if all promises are rejected.
+ */
+Q.any = any;
+
+function any(promises) {
+    var deferred = Q.defer();
```
Comment (kriskowal) [inline q.js]: This can be done after the if block to avoid allocating a deferred.

File: q.js [hunk 2]
```diff
@@ -1540,6 +1540,56 @@ Promise.prototype.all = function () {
...
+                    deferred.reject(new Error("No promises resolved successfully."));
+                }
+            }, function(progress) {
+                deferred.notify({
+                    index: i,
+                    value: progress
+                });
+            });
+        })(promise, i);
+    }
```
Comment (kriskowal) [inline q.js]: You can use array_reduce as above in Q.all, especially since you need a closure to capture the index regardless.

File: q.js [hunk 3]
```diff
@@ -1540,6 +1540,56 @@ Promise.prototype.all = function () {
...
+Q.any = any;
+
+function any(promises) {
+    var deferred = Q.defer();
+    if (promises.length === 0) {
+        return Q.resolve();
+    }
+
+    var promisesCount = promises.length;
+    for (var i = 0; i < promises.length; i++) {
```
Comment (kriskowal) [inline q.js]: Please spell out index.

File: q.js [hunk 4]
```diff
@@ -1540,6 +1540,56 @@ Promise.prototype.all = function () {
...
+            continue;
+        }
+
+        (function(promise, i) {
+            Q.resolve(promise).then(function(result) {
+                deferred.resolve(result);
+            }, function(reason) {
+                promisesCount--;
+                if(promisesCount === 0) {
+                    deferred.reject(new Error("No promises resolved successfully."));
```
Comment (kriskowal) [inline q.js]: This might be more expressive as "Can't get fulfillment value from any promise, all promises were rejected."

File: q.js [hunk 5]
```diff
@@ -1540,6 +1540,56 @@ Promise.prototype.all = function () {
...
+        /* Let's define that false could be a value to resolve with, so let's not
+         * if(!promise)
+         */
+        if (promise === undefined || promise === null) {
+            promisesCount--;
+            continue;
+        }
+
+        (function(promise, i) {
+            Q.resolve(promise).then(function(result) {
```
Comment (kriskowal) [inline q.js]: Note that Q’s internals do not generally bind to the Q object directly. Most of the internal functions are available as free variables in scope. See `when(promise, function (result) {` as in Q.all.

**Comment by kriskowal** [comment]
This looks good. Thank you for your time! If you do not have any further, I will carry this over the finish line next weekend.

**Comment by vergara** [comment]
@kriskowal 
I was not able to replace the for loop with array_reduce: reduce skips the holes in the array and since my counter depends on the length of the array, I was unable to detect undefined and decrease the counter. That made the reject for sparse array test fail.
All other comments were addressed.

**Comment by kriskowal** [comment]
@vergara Note the countDown variable in Q.all (which should be used to mirror style), that it gets incremented for each promise that array_reduce visits, then decremented for each promise that settles.

**Comment by vergara** [comment]
@kriskowal 
Updated.

File: q.js [hunk 6]
```diff
@@ -1540,6 +1540,60 @@ Promise.prototype.all = function () {
...
+        return Q.resolve();
+    }
+
+    var deferred = Q.defer();
+    var promisesCount = 0;
+    array_reduce(promises, function(prev, current, index) {
+        var promise = promises[index];
+        /* Let's define that false could be a value to resolve with, so let's not
+         * if(!promise)
+         */
```
Comment (kriskowal) [inline q.js]: // comment would suffice here.
But I think I disagree. Even `undefined` is a valid, and in fact common, promise resolution. Let’s not filter any value.

File: q.js [hunk 7]
```diff
@@ -1540,6 +1540,60 @@ Promise.prototype.all = function () {
...
+        if (promise !== undefined && promise !== null) {
+            promisesCount++;
+        } else {
+          return;
```
Comment (kriskowal) [inline q.js]: Four space indent please.

File: q.js [hunk 8]
```diff
@@ -1540,6 +1540,60 @@ Promise.prototype.all = function () {
...
+        }
+
+        (function(promise, index) {
```
Comment (kriskowal) [inline q.js]: This IIFE is no longer necessary.

File: q.js [hunk 9]
```diff
@@ -1540,6 +1540,60 @@ Promise.prototype.all = function () {
...
+            when(promise, fulfilled, rejected, progress);
+            function fulfilled(result) {
+                deferred.resolve(result);
+            };
+            function rejected(reason) {
+                promisesCount--;
+                if(promisesCount === 0) {
```
Comment (kriskowal) [inline q.js]: Space between if and paren.

File: q.js [hunk 10]
```diff
@@ -1540,6 +1540,60 @@ Promise.prototype.all = function () {
...
+ */
+Q.any = any;
+
+function any(promises) {
+    if (promises.length === 0) {
+        return Q.resolve();
+    }
+
+    var deferred = Q.defer();
+    var promisesCount = 0;
```
Comment (kriskowal) [inline q.js]: I think I may follow up by naming promisesCount and countDown above to pendingCount in both cases.

**Comment by kriskowal** [comment]
Note to self: update README and, when it lands, API reference.

**Comment by vergara** [comment]
@kriskowal 
Updated.

**Comment by kriskowal** [comment]
Thanks. Coming soon to a minor release near you.

**Pull request merged**